tests/app/views/test_suppliers.py

```diff
@@ -1749,3 +1753,261 @@ def test_changing_on_framework_to_passed_creates_audit_event(self, open_dos_fram
+        assert not json.loads(response2.get_data())["frameworkInterest"]["agreedVariations"]
+
+        # let's also check it hasn't gone & done something crazy like flipped the "alt" user/supplier's agreement flag
+        response3 = self.client.get("/suppliers/2/frameworks/g-cloud-8")
+        assert response3.status_code == 200
+        assert not json.loads(response3.get_data())["frameworkInterest"]["agreedVariations"]
+
+    def test_agree_variation_fails_with_invalid_user_id(
+            self,
+            live_g8_framework_2_variations_suppliers_on_framework,
+            ):
+        response = self.client.put("/suppliers/1/frameworks/g-cloud-8/variation/banana", data=json.dumps({
+            "updated_by": "test123",
+            "agreedVariation": {
+                "agreedUserId": 314159,
+            },
+        }), content_type="application/json")
+
+        assert response.status_code == 400
+
+        response2 = self.client.get("/suppliers/1/frameworks/g-cloud-8")
+        assert response2.status_code == 200
+        assert not json.loads(response2.get_data())["frameworkInterest"]["agreedVariations"]
+
+    def test_agree_variation_succeeds(
+            self,
+            app,
+            live_g8_framework_2_variations_suppliers_on_framework_with_alt,
+            ):
+        with freeze_time('2016-06-06'):
+            response = self.client.put("/suppliers/1/frameworks/g-cloud-8/variation/banana", data=json.dumps({
+                "updated_by": "test123",
+                "agreedVariation": {
+                    "agreedUserId": 1,
+                },
+            }), content_type="application/json")
+
+        expected_variation = {
+            "agreedUserId": 1,
+            "agreedUserEmail": "test+1@digital.gov.uk",
+            "agreedUserName": "my name",
+            "agreedAt": "2016-06-06T00:00:00.000000Z",
+        }
+
+        assert response.status_code == 200
+        response_json = json.loads(response.get_data())
+        assert response_json["agreedVariation"] == expected_variation
+
+        response2 = self.client.get("/suppliers/1/frameworks/g-cloud-8")
+        assert response2.status_code == 200
+        assert json.loads(response2.get_data())["frameworkInterest"]["agreedVariations"] == {
+            "banana": expected_variation,
+        }
+
+        with app.app_context():
+            audit_events = AuditEvent.query.filter(
+                AuditEvent.type == "agree_framework_variation"
+            ).order_by(AuditEvent.created_at, AuditEvent.id).all()
+
+            assert len(audit_events) == 1
+
+            assert audit_events[0].user == "test123"
+            assert audit_events[0].data["supplierId"] == 1
+            assert audit_events[0].data["frameworkSlug"] == "g-cloud-8"
+            assert audit_events[0].data["variationSlug"] == "banana"
+            assert audit_events[0].data["update"] == {
+                "agreedUserId": 1,
+            }
+
+    def test_agree_variation_playpen(
```

Comment:
So, yeah, this is named this because it actually tests a bunch of things together, so I found it hard to give a more specific name. And yes, in a perfect world where you have time to write a test for every possible call in every possible existing state tests would be split up, but a lot of the things tested here rely on the state built up early on in the test. The longer the test, the more rich & complex the existing state becomes, enabling the testing of more complex scenarios...

If you can think of a better name, cool. It was previously named `test_agree_variation_various`.